File: iterator-pattern/go/src/ObjectIterator.go
package src

// 对象迭代器，实现了抽象迭代器的方法，聚合了数据对象
type ObjectIterator struct {
  // 迭代器索引
  index int
  // 聚合了数据对象
  objectList *ObjectList
}

func (o *ObjectIterator) HasNext() bool {
  if o.index < o.objectList.Size() {
    return true
  }
  return false
}

func (o *ObjectIterator) Next() string {
  if o.HasNext() {
    // 返回数据对象提供的get方法，每访问一次下标增加1位
    item := o.objectList.Get(o.index)
    o.index += 1
    return item
  }
  return ""
}

File: iterator-pattern/c/src/object_iterator.c
#include "func.h"

bool has_next(Iterator *iter)
{
  if (iter->index < iter->object_list->length)
  {
    return true;
  }
  return false;
}

char *next(Iterator *iter)
{
  if (iter->has_next(iter))
  {
    // 返回数据对象提供的get方法，每访问一次下标增加1位
    char *item = iter->object_list->get(iter->object_list, iter->index);
    iter->index += 1;
    return item;
  }
  return NULL;
}

// 对象迭代器，实现了抽象迭代器的方法，聚合了数据对象
ObjectIterator *object_iterator_constructor(ObjectList *object_list)
{
  printf("\r\n object_iterator_constructor() [构建容器迭代器]");
  ObjectIterator *iter = (ObjectIterator *)malloc(sizeof(ObjectIterator));
  iter->object_list = object_list;
  iter->index = 0;
  iter->next = &next;
  iter->has_next = &has_next;
  return iter;
}